### ISSUE TITLE
MacOS TouchPad Support

### ISSUE THREAD

**Issue by PatLechevallier** [ragardner/tksheet#28]
Good Day,

https://github.com/ragardner/tksheet/blob/b1e277df52a38e8938c38fc1c220b0bf50969704/tksheet/_tksheet_main_table.py#L2222-L2232

I was investing on the handling of the touchpad scrolling because it is not working on Mac but on Windows.
I have noticed the command for ctrl-c and ctrl-v are bound to Windows ctrl key and not Mac cmd key; So I guess your library is developped for Windows.

I have investigated : https://stackoverflow.com/questions/17355902/tkinter-binding-mousewheel-to-scrollbar

> Platform differences:
> -  On Windows, you bind to <MouseWheel> and you need to divide event.delta by 120 (or some other factor depending on how fast you want the scroll)
>- on OSX, you bind to <MouseWheel> and you need to use event.delta without modification
>-  on X11 systems you need to bind to <Button-4> and <Button-5>, and you need to divide event.delta by 120 (or some other factor depending on how fast you want to scroll)

Here is my fix:

```
self.bind("<MouseWheel>", self.mousewheel)

def mousewheel(self, event=None):
        if event.state==1:
            if event.num == 5 or event.delta == -120 or event.delta == -1:
                self.MT.xview_scroll(1, "units")
                self.CH.xview_scroll(1, "units")
            if event.num == 4 or event.delta == 120 or event.delta == 1:
                if self.MT.canvasx(0) <= 0:
                    return
                self.MT.xview_scroll(-1, "units")
                self.CH.xview_scroll(-1, "units")
            self.MT.main_table_redraw_grid_and_text(redraw_row_index = True)
        else:
            if event.num == 5 or event.delta == -120 or event.delta == -1:
                self.MT.yview_scroll(1, "units")
                self.RI.yview_scroll(1, "units")
            if event.num == 4 or event.delta == 120 or event.delta == 1:
                if self.MT.canvasy(0) <= 0:
                    return
                self.MT.yview_scroll(-1, "units")
                self.RI.yview_scroll(-1, "units")
            self.MT.main_table_redraw_grid_and_text(redraw_row_index = True)
```

event.state==1 means shift is pressed during scroll so it corresponds to horizontal scrolling.
You can also do `self.bind("<Shift-MouseWheel>", self.touchpad_events)`
The fix is at the Sheet Class level, so need to remove self.MT

**Comment by ragardner** [comment]
Thanks for bringing this up,

The reasons for the current limited support for Mac - I originally just made the widget for personal use in a windows desktop app and I don't have any Mac devices to test it on

I really appreciate you taking the time to write a fix, couple of questions though,

Does your code fix all the scrolling issues for you or is there more work to be done?

Also, If it's not too much trouble, would you happen to know how I might make Copy etc work on a touch pad, does the `touchpad_events()` function receive an event after long pressing on touch screen and if so what does the object look like internally? Perhaps I can use a popup menu to allow the functionality?

Thanks for your help and testing

**Comment by ragardner** [comment]
O and lastly sorry, if possible would you be able to test if `.bind("<Mod1-` in addition to `.bind("<Control-` fixes the mac cmd key issues on a mac desktop device?

**Comment by PatLechevallier** [comment]
Re,

So I have just tested it on Windows:
It's fully working from the MouseWheel: Wheel Scrolling activates Vertical scrolling and Shift key pressed + Wheel Scrolling activate Horizontal scrolling
From the TouchPad, only the Vertical scrolling works, event with Shift pressed Horizontal scrolling doeesn't work, but it's not a code issue because I even don't receive the event.
**To Try**: Bind to `self.bind("<Shift-MouseWheel>", self.shift_mousewheel)` to see if the binding receive the event.

So for me, on Mac it works perfectly.

> Also, If it's not too much trouble, would you happen to know how I might make Copy etc work on a touch pad, does the >`touchpad_events()` function receive an event after long pressing on touch screen and if so what does the object look 
> like internally? Perhaps I can use a popup menu to allow the functionality? 

Oups, sorry `touchpad_events()` is a bad name should be just shift_mousewheel().
From what I know, there is no generic touchpad event, click, pressure and release, are just translated to basic Button events.

https://stackoverflow.com/questions/32289175/list-of-all-tkinter-events

**Comment by PatLechevallier** [comment]
Re,

I confirm `self.bind("<Command-` works.
So you it should be nice to bind depending on the OS:
https://stackoverflow.com/questions/110362/how-can-i-find-the-current-os-in-python

**Comment by ragardner** [comment]
Thanks for your help, I will let you know when I have made the updates

**Comment by PatLechevallier** [comment]
Hi,

https://github.com/ragardner/tksheet/blob/2c202864934f7fa48b356d1eb7ddd925c7f25039/tksheet/_tksheet_main_table.py#L2208-L2217

Sorry, you missed the `or event.delta == -1:` and `or event.delta == 1:`for the vertical Scroll on Mac
Beside this, it works well! You can close this ticket after the little fix.

Thank You very much!

**Comment by ragardner** [comment]
Woops, should be fixed in `4.9.3`
